### ISSUE TITLE
Any interest in doing static type checking against arcade's type hinting?

### ISSUE THREAD

**Issue by SirGnip** [pvcraven/arcade#482]
Is there any interest in taking `arcade`'s use of type hinting and going beyond "documentation" (which is great!) and starting to use it more for "checking"?  I'm talking about using static type checkers (things like "Mypy" and PyCharm's "Code -> Inspect Code") to enforce the use of arcade's type hinting.

Using static type checkers brings a number of benefits such as confirming our type hints are correct and catching code at build-time that breaks type safety.  This comes at the cost of extra discipline required by those who commit to arcade (needing to understand "type safety", which is a bit of a new frontier for some Python developers), which means that it raises the cost-of-entry to new comitters to arcade.

For grins, I ran PyCharm's "Code - Inspect Code" against the arcade repo and came up with dozens of errors and hundreds of warnings around Arcade's use of type hinting.  I also ran [`mypy`](http://mypy-lang.org/) against arcade and got similar results.  However, I'm a complete noob with `mypy` and it looked like some of the errors were related to pathing issues I have.  A number of the type checker errors PyCharm reported were simple inconsistency errors (ex: `open_window` declares `width` and `height` as `Number`, but the `aracde.Window` class declares `width` and `height` as `float`).  So, a first step would be to cleanup some of the existing type checker errors/warnings.

Thoughts?  Is this worth doing?

**Comment by pvcraven** [comment]
Yes, this is totally worth doing.

There are a few irritating spots I've not figured out how to easily get fixed. One of the most irritating is the 'must implement abstract methods' child window classes get.

**Comment by pvcraven** [comment]
Just went through and fixed a bunch of PEP-8 stuff. Type checking still needs work.

**Comment by SirGnip** [comment]
Nice.  I'll probably start slowly whittling away at some of the type checking errors as reported by PyCharm.  Once I clean up some of the low-hanging type checking fruit, I can see what possibilities there are for the "abstract method" irritation.

Then, once the codebase is pretty clean in PyCharm, we could look at getting something like `mypy` (or other linter that does Python 3 type checking) working, hooked into arcade's `make.bat/.sh`, and then hooking into CI becomes an easier possibility.  Baby steps...

**Comment by SirGnip** [comment]
A couple passes on code inspection error/warning cleanup.  This PR brought PyCharm's code inspection counts from:

36 errors, 129 warnings, 354 weak warnings

...to...

11 warnings, 165 weak warnings

...then throw in a cleanup commit or two that were merged in the last two weeks and it brings us down to:

8 warnings, 42 weak warnings

FYI... I was working with the goal "just get errors and warnings to zero" as quick as possible, so I didn't undertake any major refactoring or fixes (a number of PyCharm `noinspection` comments, etc).  Just looking to get the errors/warnings to the point where we can turn on automated code inspections.  Then, once that is in place, we can start doing the slightly larger fixes (whatever refactoring we need, etc.) with the safety net of the automated code inspection.

There are still errors/warnings that can be dealt with pretty easily, but I wanted to commit what I had before continuing to knock off the remaining.

Also, `make test` and `run_all_examples.py` ran clean after these changes.

**Comment by SirGnip** [comment]
Final commit to PR brings errors/warnings to 0 (at least in my install of PyCharm).   I just suppressed a few tricky type hinting bits instead of fixing them just to get us to 0 for now.  It is easy to go back later and search the code base for `noinspection` comments, fix the issue and remove the comments.

Now, I'll see what I can do about getting mypy to run locally.

**Comment by SirGnip** [comment]
The recent PR includes type hinting tweaks to get mypy to run cleanly.  Set your working directory to the top of the arcade repo and run `mypy -p arcade` at the top.

Most changes were minor.  Here is a summary of the changes that were a bit more "impactful":

* mypy has problems when attributes are added to function objects (see the `.cache` attribute added to the `draw_text()` function in `text.py` .  I used the `# type: ignore` comment to suppress a mypy error.  I'm not sure how to resolve this yet as I'm no mypy pro...
* `Window.__init__()` had `Number` as the type hint for width and height and `float` in its docstring.  But, pyglet has these as `int`.  I changed arcade to match pyglet as pyglet would throw an exception if a float was passed into pyglet.window.Window.__init__().
* I completely skipped type checking (`@typing.no_type_check`) on a couple deprecated functions in `read_tiled_map.py`
* Starred imports (ex `... import *`) caused mypy issues (see shader.py). This import masks a variable's origin, making mypy unable to find them and thus report errors.

**Comment by SirGnip** [comment]
@pvcraven, do you want me to go ahead and rig up type hinting to the Travis CI build so that the build fails if a commit introduces a type error?  I haven't worked with Travis CI before, but I'll give it a shot.  Also, if I do this, I think the build will be broken a lot (requires more discipline by committers).  Is this something you want?

**Comment by pvcraven** [comment]
TravisCI is broken all the time now. Mainly because of  the whole it-can't-do-graphics thing. I really wish there was a way we could run tests with OpenGL.

I'd be supportive of adding this test, but I think we should wait until the builds are back to passing.

**Comment by SirGnip** [comment]
Ok.  I'll sit tight.

Though, the commit history shows the most recent commit (d813f6e) as having a green checkmark?  Am I looking at the right thing?

**Comment by pvcraven** [comment]
I know right? I don't know for sure what fixed it. I'll have to take a look
but if it is fixed, then that'll be a great thing.

**Comment by pvcraven** [comment]
So, is this working better now?

**Comment by SirGnip** [comment]
I haven't looked at this in awhile.  The commits i did a couple months ago got the codebase clean enough to pass `mypy`.  But, I just pulled master down this afternoon and ran `make typecheck` against my local repo.  It reported 11 errors.  So, the codebase's type checking decayed (but not tons...).  I'll see if I can clean the errors up quickly. 

Also, regarding hooking type checking up to Travis, the Travis CI build appears to have been broken the last couple days with:

`ERROR: Could not find a version that satisfies the requirement pyglet-ffmpeg2 (from -r requirements.txt (line 4)) (from versions: none)
ERROR: No matching distribution found for pyglet-ffmpeg2 (from -r requirements.txt (line 4))
The command "pip3 install -r requirements.txt -r requirements_dev.txt" failed and exited with 1 during .`

**Comment by pvcraven** [comment]
Yes, even old builds are failing. pyglet-ffmpeg2 hasn't changed, so I don't know what is up with that.

**Comment by SirGnip** [comment]
I'll try throwing the mypy run in Travis CI, even if it isn't getting that far yet.  Then it is at least in git and when the Travis CI issue is resolved, we'll see how mypy does.  :)

**Comment by SirGnip** [comment]
PR is up.  I'm not familiar with TravisCI, so this PR is just a "follow the pattern" commit. 

I'm only running mypy on the Windows build so that it runs once.

I'm not sure if I can use the bash "`if` / `fi`" syntax in the .travis.yml for mypy even though I'm only running mypy on Windows?

**Comment by SirGnip** [comment]
Strangely, the TravisCI build got past the pyglet-ffmpeg2 problem for my PR.

Regarding TravisCI running mypy, I'm seeing this output:

```
$ if [ "$TRAVIS_OS_NAME" == "windows" ]; then python -m mypy -p arcade; fi
The command "if [ "$TRAVIS_OS_NAME" == "windows" ]; then python -m mypy -p arcade; fi" exited with 0.
```

I don't see any output.  But, then again that is what mypy does when it runs successfully...  So, I can't tell if it is running properly or if it isn't doing anything.  TravisCI did have a "0.00s" next to the run which makes me suspicious nothing is happening.

**Comment by pvcraven** [comment]
I think I fixed the ffmpeg thing. You can always add an error and see if it gets caught.

**Comment by SirGnip** [comment]
Heh.  I wasn't thinking.  Thanks...

I'll experiment with TravisCI and see what I can do.  I'll be pushing up a bunch of changes to this PR to poke Travis in different ways.  Don't merge the PR yet...

**Comment by SirGnip** [comment]
I've got something working.  

My first attempt had `"$TRAVIS_OS_NAME" == "windows"` because I was assuming the main build nodes were Windows nodes.  Nope.  So, mypy was never even run in my first attempt.  So, now that I'm running it against linux nodes, it is running as expected:

Example of mypy failing the build: https://travis-ci.org/pvcraven/arcade/jobs/640951146#L633
Example of mypy passing the build: https://travis-ci.org/pvcraven/arcade/jobs/640954543#L632

This PR is ready: https://github.com/pvcraven/arcade/pull/557  The current code passes the mypy run in TravisCI.

**Comment by pvcraven** [comment]
Closed with release of 2.2.7.
